src/rust/engine/src/nodes.rs

```diff
@@ -710,25 +711,46 @@ impl Node for SelectTransitive {
       .then(move |dep_product_res| {
         match dep_product_res {
           Ok(dep_product) => {
-            // The product and its dependency list are available: project them.
-            let deps =
-              future::join_all(
-                externs::project_multi(&dep_product, &self.selector.field).iter()
-                  .map(|dep_subject| self.get_dep(&context, &dep_subject))
-                  .collect::<Vec<_>>()
-              );
-            deps
-              .then(move |dep_values_res| {
-                // Finally, store the resulting values.
-                match dep_values_res {
-                  Ok(dep_values) => {
-                    Ok(self.store(&dep_product, dep_values.iter().collect()))
-                  },
-                  Err(failure) =>
-                    Err(context.was_required(failure)),
+            let addresses = externs::project_multi(&dep_product, &self.selector.field);
+
+            let init = TransitiveExpansion {
+              todo: addresses,
+              completed: HashSet::default(),
+              outputs: OrderMap::default()
+            };
+
+            loop_fn(init, move |mut expansion| {
+              let round = future::join_all({
+                 expansion.todo.drain(..)
```

Comment:
`drain` takes ownership of the items in the collection, meaning that `address` here is a `Value`, not a reference to a value (ie, `&Value`). So you should probably change the signature of `expand_transitive` to take a `Value`... that way you won't need to clone it inside that method.